### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by omugeri**

### CODE REVIEW

File: server/controllers/user.js [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        res.status(400).json({ message: 'User already exists' });
+      }
+    });
+  },
+  all: (req, res) => {
+    User.find((err, users) => {
+      if (err) {
+        return res.send(err);
+      }
+      return res.json({
```
Comment (damakahindi) [inline server/controllers/user.js:64]: There is no status returned here. I noticed you returned the status as part of the response in your above code.

File: server/controllers/user.js [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        status: 200,
+        users,
+      });
+    });
+  },
+
+  getSpecificUser: (req, res) => {
+    User.findById(req.params.user_id, (err, user) => {
+      if (err) {
+        res.status(400).json({ err });
```
Comment (damakahindi) [inline server/controllers/user.js:75]: **Consistent returns**
You need to be consistent in what the API returns. Best practice is to just return the error object.

File: coverage/coverage.json [hunk 3]
```diff
@@ -0,0 +1 @@
...e":52,"column":8}}]},"9":{"line":52,"type":"binary-expr","locations":[{"start":{"line":52,"column":12},"end":{"line":52,"column":15}},{"start":{"line":52,"column":19},"end":{"line":52,"column":24}}]},"10":{"line":59,"type":"if","locations":[{"start":{"line":59,"column":10},"end":{"line":59,"column":10}},{"start":{"line":59,"column":10},"end":{"line":59,"column":10}}]},"11":{"line":69,"type":"if","locations":[{"start":{"line":69,"column":4},"end":{"line":69,"column":4}},{"start":{"line":69,"column":4},"end":{"line":69,"column":4}}]},"12":{"line":75,"type":"if","locations":[{"start":{"line":75,"column":6},"end":{"line":75,"column":6}},{"start":{"line":75,"column":6},"end":{"line":75,"column":6}}]}}},"/Users/riwhiz/Desktop/Projects/Checkpoint_3/server/models/roles.js":{"path":"/Users/riwhiz/Desktop/Projects/Checkpoint_3/server/models/roles.js","s":{"1":1,"2":1,"3":1,"4":1},"b":{},"f":{},"fnMap":{},"statementMap":{"1":{"start":{"line":1,"column":0},"end":{"line":1,"column":37}},"2":{"start":{"line":3,"column":0},"end":{"line":3,"column":31}},"3":{"start":{"line":5,"column":0},"end":{"line":7,"column":3}},"4":{"start":{"line":9,"column":0},"end":{"line":9,"column":53}}},"branchMap":{}}}
```
Comment (damakahindi) [inline coverage/coverage.json]: You don't need the coverage folder in your GitHub repository.

File: server/models/document.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+const mongoosePaginate = require('mongoose-paginate');
+
+const Schema = mongoose.Schema;
+
+const documentSchema = new Schema({
+  Title: String,
+  Content: String,
+  Owner: String,
+  Permissions: String,
+  CreatedAt: { type: Date, default: Date.now },
```
Comment (damakahindi) [inline server/models/document.js:11]: Objects should be defined in arbnb format.
`{
type: Date,
default: Date.now(),
}`

File: server/routes/user.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  router.use((req, res, next) => {
+    // do logging
+    console.log('Something is happening. ');
+    userCntrl.authenticate(req, res);
+    next();
+  });
+
+ /*
+  //test route to make sure everything is working
+  // router.get('/', function(req, res) {
```
Comment (damakahindi) [inline server/routes/user.js:27]: Remove commented out code.

File: test/server/roles.spec.js [hunk 6]
```diff
@@ -1,10 +1,176 @@
...
+  describe('VIEW', () => {
+    it('validates that all roles are displayed', (done) => {
+      request
+        .get('/api/roles/')
+        .set({ 'x-access-token': token })
+        .end((err, res) => {
+          if (err) {
+            return done(err);
+          }
+          console.log('RES IS: ', res.body[4]);
```
Comment (damakahindi) [inline test/server/roles.spec.js]: Log statements on your code.

File: test/server/users.spec.js [hunk 7]
```diff
@@ -1,18 +1,298 @@
...
+        email: 'laycee@gmail.com',
+      })
+      .end((err, res) => {
+        if (err) {
+          return done(err);
+        }
+        expect(res.status).to.be.equal(400);
+        expect(res.body.message).to.exist;
+        expect(res.body.message).to.be.equal('Unauthorized User!')
+        // expect(res.body.name).to.not.exist;
```
Comment (damakahindi) [inline test/server/users.spec.js]: Commented out code

File: server.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+mongoose.Promise = global.Promise;
+mongoose.connect('mongodb://localhost/docman');
+
+// configure app to use bodyParser()
+// this will let us get data from a POST
+app.use(morgan('dev')); // log every request to the console
+app.use(cookieParser(process.env.SECRET)); // read cookieParser
+app.use(bodyParser.urlencoded({ extended: true }));
+app.use(bodyParser.json());
+app.use(session({
```
Comment (kndungu) [inline server.js]: Unnecessary

File: server.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  secret: process.env.SECRET,
+  resave: true,
+  saveUninitialized: true,
+}));
+app.use(passport.initialize());
```
Comment (kndungu) [inline server.js]: Unnecessary

File: server.js [hunk 10]
```diff
@@ -0,0 +1,47 @@
+// call the packages we need
```
Comment (kndungu) [inline server.js:1]: Passport and sessions are unnecessary

File: server/models/user.js [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    last: { type: String, required: true },
+  },
+  email: { type: String, required: true },
+  password: String,
+  role: String,
+});
+
+// generate a hash
+userSchema.methods.generateHash = (password) => {
+  return bcrypt.hashSync(password, bcrypt.genSaltSync(8), null);
```
Comment (kndungu) [inline server/models/user.js:19]: Using the asynchronous version of function will make your backend faster

File: server/controllers/user.js [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+    const token = req.body.token || req.query.token || req.headers['x-access-token'];
+    if (token) {
+      let verified = nJwt.verify(token, secret);
+      if (verified === { JwtParseError: 'Jwt cannot be parsed' }) {
+        verified = { message: 'Unauthorized User!' };
+        return verified;
+      }
+      return verified;
+    }
+    const verified = { message: 'Unauthorized User!' };
```
Comment (kndungu) [inline server/controllers/user.js]: This assignment is unnecessary.

File: server/controllers/user.js [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+
+        // save the user and check for errors
+        user.save(() => {
+          if (err) {
+            res.status(400).json({
+              message: 'Something went wrong',
+              err,
+            });
+          } else {
+            res.status(200).json({ message: 'User Created', user });
```
Comment (kndungu) [inline server/controllers/user.js]: __Best Practice__

Only return the user, the status code lets the client know the action was successful

File: server/controllers/user.js [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+  },
+
+  deleteUser: (req, res) => {
+    User.remove({
+      _id: req.params.user_id,
+    }, (err) => {
+      if (err) {
+        return res.status(400).json({ message: 'Unsuccessfull', err });
+      }
+      return res.json({ message: 'Successfully deleted' });
```
Comment (kndungu) [inline server/controllers/user.js:120]: __Best Practice__

Return the deleted object

File: test/server/users.spec.js [hunk 15]
```diff
@@ -1,18 +1,295 @@
...
+const request = require('supertest')(app);
+const expect = require('chai').expect;
+
+
+describe('User', () => {
+  let token;
+  let id;
+  let id2;
+  const wrongId = '57df94dfe74e460dead6792';
+  beforeEach((done) => {
```
Comment (kndungu) [inline test/server/users.spec.js]: `before` would work just as well without the additional processing over head

**Review by kndungu — COMMENTED** [review]
And in general comment your code.

**Pull request merged**